Replace `ExecuteSelectionSet` with `ExecuteGroupedFieldSet` (#1039)
graphql/graphql-spec: spec/Section 6 -- Execution.md

- Let {mutationType} be the root Mutation type in {schema}.
- Assert: {mutationType} is an Object type.
- Let {selectionSet} be the top level selection set in {mutation}.
- Return {ExecuteRootSelectionSet(variableValues, initialValue, mutationType,

>   selectionSet, true)}.

Reviewer: I don't believe a trailing optional boolean argument is very clear, an enum of strings might be easier to read

```graphql example
{
  a {
    subfield1
  }
  ...ExampleFragment
}

fragment ExampleFragment on Query {
  a {
    subfield2
  }
  b
}
```

The depth-first-search order of the field groups produced by {CollectFields()}
is maintained through execution, ensuring that fields appear in the executed
response in a stable and predictable order.

CollectFields(objectType, selectionSet, variableValues, visitedFragments):

- If {visitedFragments} is not provided, initialize it to the empty set.
- Initialize {groupedFields} to an empty ordered map of lists.
- For each {selection} in {selectionSet}:
  - If {selection} provides the directive `@skip`, let {skipDirective} be that
    directive.
    - If {skipDirective}'s {if} argument is {true} or is a variable in
      {variableValues} with the value {true}, continue with the next {selection}
      in {selectionSet}.
  - If {selection} provides the directive `@include`, let {includeDirective} be
    that directive.
    - If {includeDirective}'s {if} argument is not {true} and is not a variable
      in {variableValues} with the value {true}, continue with the next
      {selection} in {selectionSet}.
  - If {selection} is a {Field}:

>     - Let {responseKey} be the response key of {selection} (the alias if

Reviewer 2: ```suggestion
    - Let {responseName} be the response name of {selection} (the alias if
```

      defined, otherwise the field name).

>     - Let {groupForResponseKey} be the list in {groupedFields} for

Reviewer 2: ```suggestion
    - Let {groupForResponseName} be the list in {groupedFields} for
```

>       {responseKey}; if no such list exists, create it as an empty list.

Reviewer 2: ```suggestion
      {responseName}; if no such list exists, create it as an empty list.
```

>     - Append {selection} to the {groupForResponseKey}.

Reviewer 2: ```suggestion
    - Append {selection} to the {groupForResponseName}.
```

  - If {selection} is a {FragmentSpread}:
    - Let {fragmentSpreadName} be the name of {selection}.
    - If {fragmentSpreadName} is in {visitedFragments}, continue with the next
      {selection} in {selectionSet}.
    - Add {fragmentSpreadName} to {visitedFragments}.
    - Let {fragment} be the Fragment in the current Document whose name is
      {fragmentSpreadName}.
    - If no such {fragment} exists, continue with the next {selection} in
      {selectionSet}.
    - Let {fragmentType} be the type condition on {fragment}.
    - If {DoesFragmentTypeApply(objectType, fragmentType)} is {false}, continue
      with the next {selection} in {selectionSet}.
    - Let {fragmentSelectionSet} be the top-level selection set of {fragment}.
    - Let {fragmentGroupedFieldSet} be the result of calling
      {CollectFields(objectType, fragmentSelectionSet, variableValues,
      visitedFragments)}.
    - For each {fragmentGroup} in {fragmentGroupedFieldSet}:

>       - Let {responseKey} be the response key shared by all fields in

Reviewer 2: ```suggestion
      - Let {responseName} be the response name shared by all fields in
```

        {fragmentGroup}.
      - Let {groupForResponseKey} be the list in {groupedFields} for

>         {responseKey}; if no such list exists, create it as an empty list.

Reviewer 2: ```suggestion
        {responseName}; if no such list exists, create it as an empty list.
```

>       - Append all items in {fragmentGroup} to {groupForResponseKey}.

Reviewer 2: ```suggestion
      - Append all items in {fragmentGroup} to {groupForResponseName}.
```

  - If {selection} is an {InlineFragment}:
    - Let {fragmentType} be the type condition on {selection}.
    - If {fragmentType} is not {null} and {DoesFragmentTypeApply(objectType,
      fragmentType)} is {false}, continue with the next {selection} in
      {selectionSet}.
    - Let {fragmentSelectionSet} be the top-level selection set of {selection}.
    - Let {fragmentGroupedFieldSet} be the result of calling
      {CollectFields(objectType, fragmentSelectionSet, variableValues,
      visitedFragments)}.
    - For each {fragmentGroup} in {fragmentGroupedFieldSet}:

>       - Let {responseKey} be the response key shared by all fields in

Reviewer 2: ```suggestion
      - Let {responseName} be the response name shared by all fields in
```

        {fragmentGroup}.

>       - Let {groupForResponseKey} be the list in {groupedFields} for
>         {responseKey}; if no such list exists, create it as an empty list.
>       - Append all items in {fragmentGroup} to {groupForResponseKey}.

Reviewer 2: ```suggestion
      - Let {groupForResponseName} be the list in {groupedFields} for
        {responseName}; if no such list exists, create it as an empty list.
      - Append all items in {fragmentGroup} to {groupForResponseName}.
```

- Cancel {responseStream}.

> ## Executing the Root Selection Set

Reviewer: this new header name seems like it should not be at `##` level anymore, but the sub-algos still matter. 

Perhaps this top level section should remain and explain the high level idea, and then have these be sub-headers

the underlying Source Stream).

First, the selection set is turned into a _grouped field set_; then, we execute
this grouped field set and return the resulting {data} and {errors}.

ExecuteRootSelectionSet(variableValues, initialValue, objectType, selectionSet,
serial):

- If {serial} is not provided, initialize it to {false}.
- Let {groupedFieldSet} be the result of {CollectFields(objectType,
  selectionSet, variableValues)}.
- Let {data} be the result of running {ExecuteGroupedFieldSet(groupedFieldSet,
  objectType, initialValue, variableValues)} _serially_ if {serial} is {true},
  _normally_ (allowing parallelization) otherwise.
- Let {errors} be the list of all _execution error_ raised while executing the
  selection set.
- Return an unordered map containing {data} and {errors}.

### Field Collection

:: A _grouped field set_ is a map where each entry is a list of field selections
that share a _response name_ (the alias if defined, otherwise the field name).

Before execution, the _selection set_ is converted to a _grouped field set_ by
calling {CollectFields()}. This ensures all fields with the same response name

> (including those in referenced fragments) are executed at the same time.

Author: These are the only lines that have actually been changed in Field Collection, everything else was just moved.

(To prove this, take the current `Field Collection` text and paste it over this PR, the diff should render as:

```
diff --git i/spec/Section 6 -- Execution.md w/spec/Section 6 -- Execution.md
index f81e38d..549930f 100644
--- i/spec/Section 6 -- Execution.md	
+++ w/spec/Section 6 -- Execution.md	
@@ -374,12 +374,11 @@ serial):
 
 ### Field Collection
 
-:: A _grouped field set_ is a map where each entry is a list of field selections
-that share a _response name_ (the alias if defined, otherwise the field name).
-
-Before execution, the _selection set_ is converted to a _grouped field set_ by
-calling {CollectFields()}. This ensures all fields with the same response name
-(including those in referenced fragments) are executed at the same time.
+Before execution, the _selection set_ is converted to a grouped field set by
+calling {CollectFields()}. Each entry in the grouped field set is a list of
+fields that share a _response name_ (the alias if defined, otherwise the field
+name). This ensures all fields with the same response name (including those in
+referenced fragments) are executed at the same time.
 
 As an example, collecting the fields of this selection set would collect two
 instances of the field `a` and one of field `b`:
```
)

process continues until there are no more subfields to collect and execute.

### Executing the Root Selection Set

:: A _root selection set_ is the top level _selection set_ provided by a GraphQL
operation. A root selection set always selects from a root type.

To execute the root selection set, the initial value being evaluated and the
root type must be known, as well as whether it must be executed serially, or may
be executed in parallel (see
[Normal and Serial Execution](#sec-Normal-and-Serial-Execution).

Executing the root selection set works similarly for queries (parallel),
mutations (serial), and subscriptions (where it is executed for each event in
the underlying Source Stream).

First, the _selection set_ is collected into a _grouped field set_ which is then
executed, returning the resulting {data} and {errors}.

ExecuteRootSelectionSet(variableValues, initialValue, objectType, selectionSet,
executionMode):

- Let {groupedFieldSet} be the result of {CollectFields(objectType,
  selectionSet, variableValues)}.
- Let {data} be the result of running {ExecuteGroupedFieldSet(groupedFieldSet,
  objectType, initialValue, variableValues)} _serially_ if {executionMode} is

>   {"serial"}, otherwise _normally_).

Author: I think we should restore the parenthesis, they add clarity to what "normally" means.

```suggestion
  {"serial"}, otherwise _normally_ (allowing parallelization)).
```

- Let {errors} be the list of all _execution error_ raised while executing the
  selection set.
- Return an unordered map containing {data} and {errors}.

### Field Collection

Before execution, the _selection set_ is converted to a _grouped field set_ by
calling {CollectFields()}. This ensures all fields with the same response name,
including those in referenced fragments, are executed at the same time.

> :: A _grouped field set_ is a map where each entry is a _response name_ and a
> list of selected fields that share that _response name_ (the field alias if
> defined, otherwise the field's name).

Author: If we choose not to rename "grouped field set", we should at least define it in terms of a set:

```suggestion
:: A _grouped field set_ is a map where each entry is a _response name_ and an
ordered set of selected fields that share that _response name_ (the field alias
if defined, otherwise the field's name).
```

Author: If we accept this, we should also make a few other changes such as:

```diff
-Initialize groupedFields to an empty ordered map of lists.
+Initialize groupedFields to an empty ordered map of ordered sets.
```

Reviewer 3: There's no functional change between a list or a set in this case since the algorithm never attempts to add the same item to the set/list?

Author: Yes indeed, list or ordered set is equivalent here since it will never include duplicates. The main reason to change it to set is because we call it a "grouped field set"; a "field set" being `OrderedSet<Field>`, and grouped being a map grouped by response key: `Map<string, OrderedSet<Field>>`.

Author: This might be a simpler solve actually:

```suggestion
:: A _grouped field set_ is a map where each entry is a _response name_ and a
list of selected fields that share that _response name_ (the field alias if
defined, otherwise the field's name).

Note: No entry in the list of selected fields will be duplicated, hence we call
it a "field set" even though in the algorithms we use a list structure for it.
```

      {CollectFields(objectType, fragmentSelectionSet, variableValues,
      visitedFragments)}.
    - For each {fragmentGroup} in {fragmentGroupedFieldSet}:
      - Let {responseName} be the response name shared by all fields in
        {fragmentGroup}.
      - Let {groupForResponseName} be the list in {groupedFields} for
        {responseName}; if no such list exists, create it as an empty list.
      - Append all items in {fragmentGroup} to {groupForResponseName}.
- Return {groupedFields}.

DoesFragmentTypeApply(objectType, fragmentType):

- If {fragmentType} is an Object Type:
  - If {objectType} and {fragmentType} are the same type, return {true},
    otherwise return {false}.
- If {fragmentType} is an Interface Type:
  - If {objectType} is an implementation of {fragmentType}, return {true}
    otherwise return {false}.
- If {fragmentType} is a Union:
  - If {objectType} is a possible type of {fragmentType}, return {true}
    otherwise return {false}.

Note: The steps in {CollectFields()} evaluating the `@skip` and `@include`
directives may be applied in either order since they apply commutatively.

**Merging Selection Sets**

> When more than one field of the same name is executed in parallel, during value
> completion each related _selection set_ is collected together to produce a
> single _grouped field set_ in order to continue execution of the sub-selection
> sets.

Author: Text like this already exists in the spec as it stands today, but it is confusing. The "in parallel" sounds like it applies to serial versus parallel execution, but it doesn't since this still holds for mutation fields. Further the "same name" is also not quite right, because it's the _response name_ that matters, not the actual field name - `{ cat: pet(id: 1) { name } dog: pet(id: 2) { age } }` would not merge selection sets.

The thing is, we don't execute _a field_ (in the GraphQL request document AST sense[^1] - i.e. a "field selection") we actually execute _a field set_ (i.e. a "set of field selections").

[^1]: We do execute a field in the GraphQL schema sense. Using "field" for both of these is quite confusing, I find. Maybe we should be more careful to use "field selection" rather than "field".

How about:

```suggestion
When a field is executed, during value completion the _selection set_ of each of
the related field selections with the same response name are collected together
to produce a single _grouped field set_ in order to continue execution of the
sub-selection sets.
```

  objectType, initialValue, variableValues)} _serially_ if {executionMode} is
  {"serial"}, otherwise _normally_ (allowing parallelization)).
- Let {errors} be the list of all _execution error_ raised while executing the
  selection set.
- Return an unordered map containing {data} and {errors}.

### Field Collection

Before execution, the _root selection set_ is converted to a _grouped field set_
by calling {CollectFields()}. This ensures all fields with the same response
name, including those in referenced fragments, are executed at the same time.

:: A _grouped field set_ is a map where each entry is a _response name_ and its
associated _field set_. A _grouped field set_ may be produced from a selection
set via {CollectFields()} or from the selection sets of a _field set_ via
{CollectSubfields()}.

:: A _field set_ is a list of selected fields that share the same _response
name_ (the field alias if defined, otherwise the field's name).

Note: The order of field selections in a _field set_ is significant, hence the
algorithms in this specification model it as a list. Any later duplicated field
selections in a field set will not impact its interpretation, so using an

> ordered set would yield equivalent results.

Author: # :new:

After resolving the value for `me`, the selection sets are merged together so
`firstName` and `lastName` can be resolved for one value.

MergeSelectionSets(fields):
CollectSubfields(objectType, fields, variableValues):

- Let {selectionSet} be an empty list.
- Let {groupedFieldSet} be an empty map.
- For each {field} in {fields}:
  - Let {fieldSelectionSet} be the selection set of {field}.
  - If {fieldSelectionSet} is null or empty, continue to the next field.
  - Append all selections in {fieldSelectionSet} to {selectionSet}.

> - Return {selectionSet}.

Author: Perhaps `childGroupedFieldSet`? `fieldGroupedFieldSet` sounds to me like a "field-grouped field set" rather than a "field's grouped-field-set".

> - Return {selectionSet}.

Reviewer 4: Hmm, child has the same problem, but maybe if we change fieldSelectionSet in tandem that would work anyway through some context signaling

After resolving the value for `me`, the selection sets are merged together so
`firstName` and `lastName` can be resolved for one value.

CollectSubfields(objectType, fields, variableValues):

- Let {groupedFieldSet} be an empty map.
- For each {field} in {fields}:
  - Let {fieldSelectionSet} be the selection set of {field}.
  - If {fieldSelectionSet} is null or empty, continue to the next field.
  - Let {fieldGroupedFieldSet} be the result of {CollectFields(objectType,
    fieldSelectionSet, variableValues)}.

>   - For each {fieldGroupedFieldSet} as {responseKey} and {subfields}:
>     - Let {groupForResponseKey} be the list in {groupedFieldSet} for
>       {responseKey}; if no such list exists, create it as an empty list.
>     - Append all fields in {subfields} to {groupForResponseKey}.
> - Return {groupedFieldSet}.
> 
> Note: All the {fields} passed to {CollectSubfields()} share the same _response
> key_.

Reviewer 2: ```suggestion
  - For each {fieldGroupedFieldSet} as {responseKey} and {subfields}:
    - Let {groupForResponseName} be the list in {groupedFieldSet} for
      {responseName}; if no such list exists, create it as an empty list.
    - Append all fields in {subfields} to {groupForResponseName}.
- Return {groupedFieldSet}.

Note: All the {fields} passed to {CollectSubfields()} share the same _response
name_.
```
